### DISCUSSION TITLE
using java Runtime.getRuntime().availableProcessors() for channel count config

### DISCUSSION THREAD

**Discussion by guykatz** [microstream-one/microstream#562] [Q&A] [answered]
hi all;
is it a good idea to use  int processors = Runtime.getRuntime().availableProcessors(); to determine, dynamically, the channel count in a microstream storage such that it is optimized to the platform it runs on?
thanks

**Comment by fh-ms** [comment, marked as answer]
Hi, this is basically a good idea. Keep in mind that the number of channels must be a power of 2.
An example can be found in the Bookstore Demo:

https://github.com/microstream-one/bookstore-demo/blob/master/src/main/java/one/microstream/demo/bookstore/BookStoreDemo.java#L165

It doesn't utilize all cores for the channels, in order to leave some for the OS, housekeeping and the application itself.

**Comment by hg-ms** [comment]
I have some doubts about using a dynamic set channel count:

1. One a storage has been started the channel count can’t be changed on the fly. To do so the storage must be converted explicitly.
2. On bigger machines there might be much more channels created then required.

3. In the past Runtime.getRuntime().availableProcessors() reported unexpected values on some systems. I am not sure how it behaves currently.

**Reply by guykatz** [reply]
and given [https://github.com/microstream-one/microstream/discussions/566](https://github.com/microstream-one/microstream/discussions/566)
I think this might blow up in many people faces... imagine even a simple channel count tuning between product version that is delivered to hundreds of customers...
we need to be very careful using this I believe
